File: tests/tooling/perf/matmul.c
// Dense square matrix multiply over flat double arrays: the classic
// triple loop, indexed with i*n+k arithmetic rather than a counter the
// compiler can bound. Nothing allocates after setup, so it isolates
// float FMA-less arithmetic and strided loads from everything else.
#include <stdio.h>
#include <stdlib.h>

#define N 320

int main(void) {
    double *a = malloc(sizeof(double) * N * N);
    double *b = malloc(sizeof(double) * N * N);
    double *c = malloc(sizeof(double) * N * N);
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++) {
            a[i * N + j] = (double)((i + j) % 7);
            b[i * N + j] = (double)((i * 2 + j) % 5);
            c[i * N + j] = 0.0;
        }
    for (int rep = 0; rep < 3; rep++)
        for (int i = 0; i < N; i++)
            for (int k = 0; k < N; k++) {
                double aik = a[i * N + k];
                for (int j = 0; j < N; j++)
                    c[i * N + j] += aik * b[k * N + j];
            }
    double s = 0.0;
    for (int i = 0; i < N * N; i++) s += c[i];
    printf("%.0f\n", s);
    return 0;
}

File: tests/tooling/perf/matmul.rs
// The Rust twin of matmul.fpp: the classic triple loop over flat arrays,
// indexed with i*n+k arithmetic. Bounds checks are ON — this is what a
// checked flat-matrix index costs.
const N: usize = 320;

fn main() {
    let mut a = vec![0.0f64; N * N];
    let mut b = vec![0.0f64; N * N];
    let mut c = vec![0.0f64; N * N];
    for i in 0..N {
        for j in 0..N {
            a[i * N + j] = ((i + j) % 7) as f64;
            b[i * N + j] = ((i * 2 + j) % 5) as f64;
        }
    }
    for _ in 0..3 {
        for i in 0..N {
            for k in 0..N {
                let aik = a[i * N + k];
                for j in 0..N { c[i * N + j] += aik * b[k * N + j]; }
            }
        }
    }
    let mut s = 0.0f64;
    for i in 0..N * N { s += c[i]; }
    println!("{:.0}", s);
}
